Move renderers from flow-server to flow-data. Rework the ComponentRenderer class hierarchy and internal implementations.

gilberto-torrezan opened the pull request.

denis-anisimov commented:
Reviewed 33 of 33 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/ComponentDataGenerator.java, line 144 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o2qUxJzJZ6MQjgsgC:-L4o2qUxJzJZ6MQjgsgD:bz24fm8) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/provider/ComponentDataGenerator.java#L144)):*
> public void setComponentRenderer(
>             ComponentRenderer<? extends Component, T> componentRenderer) {
>         this.componentRenderer = componentRenderer;
>     }

Is there any usecase where the setters are really needed?
What's the purpose of changing `final` fields to be not `final` ?

Setters (especially `public`) defines API which you _HAVE_ to support.
It means that in many many cases changes the value via setter should report about this change (e.g. via listeners) to anyone who relies on the current values.
Changing the field via setter changes the internal state of the object and it's quite often that this change should be propagated to anyone who uses the class relying on its internal state.

Here nothing is done to report about state change.
I don't know the whole logic of this PR: may be there is no need to reset clients of this class when state has changed but normally there is such need.

That's the reason why you should _ALWAYS_ think many many times when you introduces setters.
It's quite easy to introduce it and it seems it doesn't break/change anything.
In reality you are breaking encapsulation.

It's always better to construct class once via parameters in its CTOR (so that they are final) and don't care in the future about possible state change (even though the CTOR may looks ugly). 

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 48 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5-nY5z0jBI4LWUBX:-L4o5-nY5z0jBI4LWUBY:bojy96z) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L48)):*
> private ValueProvider<SOURCE, TARGET> valueProvider;

`final` I think (related to  my comment above about setter).

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 70 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5HTp-Ah76kJpimm8:-L4o5HTp-Ah76kJpimm9:bn5qzj5) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L70)):*
> protected BasicRenderer() {
>     }

I don't see a setter for `valueProvider` here. How do you deal with `valueProvider` which is `null` ?

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 167 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5vcR8eH8Iaq6DIl3:-L4o5vcR8eH8Iaq6DIl4:bmuoy20) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L167)):*
> valueProvider

In the default CTOR it's `null`  which means NPE here.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o6wjzkwk4QIILqJSL:-L4o6wk-yEJtW8TO0teP:b1bhzni) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L55)):*
> private SerializableSupplier<COMPONENT> componentSupplier;
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     private SerializableFunction<SOURCE, COMPONENT> componentFunction;
>     private SerializableBiConsumer<COMPONENT, SOURCE> itemConsumer;
>     private Map<String, String> rendererAttributes = new HashMap<>();
>     private String componentRendererTag = "flow-component-renderer";
</details>

some of them can be `final` ?

(especially  `rendererAttributes` )

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 119 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o6IORYH3XpFYVOJuV:-L4o6IORYH3XpFYVOJuW:b-glycb4) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L119)):*
> protected ComponentRenderer() {
>     }

Any other CTOR can be used by subclasses.
The only question is which values are absolutely necessary to be able to construct this class so that it functioning appropriately.

I see this CTOR as an evil at the moment.
At least the javadoc is confusing.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 180 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o7EGwIPpHNBpZW45j:-L4o7EGwIPpHNBpZW45k:b-3upvgy) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L180)):*
> public String getComponentRendererTag() {

Is it really needed as public API ?

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/NativeButtonRenderer.java, line 56 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o7qSW-6G39s9dqXVr:-L4o7qSW-6G39s9dqXVs:b-i5iofh) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/NativeButtonRenderer.java#L56)):*
> this(SOURCE -> label);

?

Does `SOURCE` hide `SOURCE` type variable name ? (they can't be the same, right ? since here it's the formal parameter name).

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java, line 56 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o95SVetrmIXrG9VZR:-L4o95SWisOrSbFDNHXH:b-bwuofl) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java#L56)):*
> private String template;
>     private Map<String, ValueProvider<SOURCE, ?>> valueProviders;
>     private Map<String, SerializableConsumer<SOURCE>> eventHandlers;

Some of the may be `final`  ?

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o8PwWURFyK013B5_S:-L4o8PwWURFyK013B5_T:bdkc6ig) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java#L64)):*
> valueProviders = new HashMap<>(0);

`0` as the initial capacity ?
It means that once the very first item (actually the second) is added the map will require restructure.
Do you expect only one item in most cases ?
Why don't use the default capacity ?
Or if you care about memory consumption then it should just use lazy loading.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 16 of 22 files reviewed at latest revision, 10 unresolved discussions.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/ComponentDataGenerator.java, line 144 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o2qUxJzJZ6MQjgsgC:-L4oFps4166tbTsI8qLG:b-qvrqm6) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/provider/ComponentDataGenerator.java#L144)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> public void setComponentRenderer(
>             ComponentRenderer<? extends Component, T> componentRenderer) {
>         this.componentRenderer = componentRenderer;
>     }

Is there any usecase where the setters are really needed?
What's the purpose of changing `final` fields to be not `final` ?

Setters (especially `public`) defines API which you _HAVE_ to support.
It means that in many many cases changes the value via setter should report about this change (e.g. via listeners) to anyone who relies on the current values.
Changing the field via setter changes the internal state of the object and it's quite often that this change should be propagated to anyone who uses the class relying on its internal state.

Here nothing is done to report about state change.
I don't know the whole logic of this PR: may be there is no need to reset clients of this class when state has changed but normally there is such need.

That's the reason why you should _ALWAYS_ think many many times when you introduces setters.
It's quite easy to introduce it and it seems it doesn't break/change anything.
In reality you are breaking encapsulation.

It's always better to construct class once via parameters in its CTOR (so that they are final) and don't care in the future about possible state change (even though the CTOR may looks ugly). 
</blockquote></details>

Refactored the hierarchy of the ComponentDataGenerator. Added an AbstractComponentGenerator which is now the parent class of the `SimpleValueRendering` (not the `ComponentDataGenerator` directly anymore).

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 48 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5-nY5z0jBI4LWUBX:-L4od6T0MYZxOrBA6C4E:byg433c) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L48)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private ValueProvider<SOURCE, TARGET> valueProvider;

`final` I think (related to  my comment above about setter).
</blockquote></details>

Refactored.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 70 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5HTp-Ah76kJpimm8:-L4od93bfSYLJnSVFvP2:byg433c) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L70)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> protected BasicRenderer() {
>     }

I don't see a setter for `valueProvider` here. How do you deal with `valueProvider` which is `null` ?
</blockquote></details>

Refactored.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java, line 167 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o5vcR8eH8Iaq6DIl3:-L4odg6tmD0mnfbmLRe2:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/BasicRenderer.java#L167)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> valueProvider

In the default CTOR it's `null`  which means NPE here.
</blockquote></details>

Done.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 55 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o6wjzkwk4QIILqJSL:-L4oe4bz8CfNhXCrGE8O:bn0w693) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L55)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private SerializableSupplier<COMPONENT> componentSupplier;
>     private SerializableFunction<SOURCE, COMPONENT> componentFunction;
>     private SerializableBiConsumer<COMPONENT, SOURCE> itemConsumer;
>     private Map<String, String> rendererAttributes = new HashMap<>();
>     private String componentRendererTag = "flow-component-renderer";

some of them can be `final` ?

(especially  `rendererAttributes` )
</blockquote></details>

Actually the `rendererAttributes` is not needed at all. Removed.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 119 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o6IORYH3XpFYVOJuV:-L4oeKGAp_E3F_vw__q1:b-13tex2) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L119)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> protected ComponentRenderer() {
>     }

Any other CTOR can be used by subclasses.
The only question is which values are absolutely necessary to be able to construct this class so that it functioning appropriately.

I see this CTOR as an evil at the moment.
At least the javadoc is confusing.
</blockquote></details>

`BasicRenderer`, which extends `ComponentRenderer`, actually uses a different approach to the component generation (it takes a ValueProvider instead of a Component function/supplier). That makes it harder to use the default CTORS of the `ComponentRenderer`. This, for instance, is not valid:

```
protected BasicRenderer(ValueProvider<SOURCE, TARGET> valueProvider) {
    super(item -> new Span(getFormattedValue(valueProvider.apply(item))));
}
```
... because `getFormattedValue` is a method of the instance.

The proper way to solve this is to extract yet another class from `ComponentRenderer`, and make `BasicRenderer` extend from it instead. 

But then it creates some sort of a problem with the original design: components are supposed to have a `setRenderer(ComponentRenderer)` method (forcing all renderers that support components to be subclasses of `ComponentRenderer`), and the end user is supposed to use the API by invoking `component.setRenderer(new ComponentRenderer(...))`, which forces the `ComponentRenderer`  class to be a concrete one.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java, line 180 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o7EGwIPpHNBpZW45j:-L4ogoIHVQ-oy9BxyqkB:bss4a2b) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/ComponentRenderer.java#L180)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> public String getComponentRendererTag() {

Is it really needed as public API ?
</blockquote></details>

Removed.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/NativeButtonRenderer.java, line 56 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o7qSW-6G39s9dqXVr:-L4oiKT2r3lXCr0s40AJ:b-9gsd0a) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/NativeButtonRenderer.java#L56)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> this(SOURCE -> label);

?

Does `SOURCE` hide `SOURCE` type variable name ? (they can't be the same, right ? since here it's the formal parameter name).
</blockquote></details>

It doesn't hide, but it is weird anyway. Fixed.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java, line 56 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o95SVetrmIXrG9VZR:-L4oifJVz56NeAjM4nQ4:bcl1hjo) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java#L56)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private String template;
>     private Map<String, ValueProvider<SOURCE, ?>> valueProviders;
>     private Map<String, SerializableConsumer<SOURCE>> eventHandlers;

Some of the may be `final`  ?
</blockquote></details>

Not anymore, since the variables are now lazy loaded.

---

*[flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java, line 64 at r1](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4o8PwWURFyK013B5_S:-L4okEgF8DkXjgbsUee2:bvriqze) ([raw file](https://github.com/vaadin/flow/blob/30435ed6dfbc6c971705171b04cc9ddf3d10e678/flow-data/src/main/java/com/vaadin/flow/data/renderer/Renderer.java#L64)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> valueProviders = new HashMap<>(0);

`0` as the initial capacity ?
It means that once the very first item (actually the second) is added the map will require restructure.
Do you expect only one item in most cases ?
Why don't use the default capacity ?
Or if you care about memory consumption then it should just use lazy loading.
</blockquote></details>

Lazy loading it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 16 of 22 files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4olDOn5TaYRHmZrCWV:-L4olDOn5TaYRHmZrCWW:b-8pzo5b) (no related file):*
Don't merge this PR until the `alpha20` is released.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 33 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 36 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omVzBmfrXJtb73pnM:-L4omVzBmfrXJtb73pnN:b-crmb5) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L36)):*
> private Map<String, Component> renderedComponents;

`final` ?

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 87 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omoMM0EeA_B_b9OUO:-L4omoMM0EeA_B_b9OUP:b6tlw4z) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L87)):*
> this.container = container;

And again the question about this setter: shouldn't the components be rerendered (or whatever) if the container is changed ?
Or it's intended to be used once ?

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 122 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omgntHS9C8QbIaRVO:-L4omgnukHUjTITAkHIy:b69ujqe) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L122)):*
>  container.

Technically should be access via the getter since it's overridable. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: 18 of 22 files reviewed at latest revision, 4 unresolved discussions.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 36 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omVzBmfrXJtb73pnM:-L4tTez4OnUBlhYQGCxR:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L36)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> private Map<String, Component> renderedComponents;

`final` ?
</blockquote></details>

Done.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 87 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omoMM0EeA_B_b9OUO:-L4tTg38TJw3ipv6Clh-:b-rvk8vw) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L87)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

> this.container = container;

And again the question about this setter: shouldn't the components be rerendered (or whatever) if the container is changed ?
Or it's intended to be used once ?
</blockquote></details>

Refactored. Now the subclasses are responsible for managing the container.

---

*[flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java, line 122 at r2](https://reviewable.io:443/reviews/vaadin/flow/3507#-L4omgntHS9C8QbIaRVO:-L4tTm1r_PCd6cHZNzTN:b-896fix) ([raw file](https://github.com/vaadin/flow/blob/ba9a6f16ccc39ff6b4b0ae71556375b55f83cacf/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java#L122)):*
<details><summary><i>Previously, denis-anisimov (Denis) wrote…</i></summary><blockquote>

>  container.

Technically should be access via the getter since it's overridable. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 2 of 33 files at r1, 2 of 2 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 23 unresolved discussions, some commit checks broke.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->

denis-anisimov commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/flow/3507)*
<!-- Sent from Reviewable.io -->
